python/pyspark/sql/pandas/serializers.py

```diff
@@ -169,6 +169,8 @@ def create_array(s, t):
             elif is_categorical_dtype(s.dtype):
                 # Note: This can be removed once minimum pyarrow version is >= 0.16.1
                 s = s.astype(s.dtypes.categories.dtype)
+            elif t is not None and pa.types.is_string(t):
+                s = s.astype(str)
```

Comment:
so are you saying that the type of strings from pandas can produce different type in arrow? cc @BryanCutler

Reply:
while I understand that we can work around,

> Pandas stores string columns in two different ways: using a numpy `ndarray` or using a custom `StringArray`. The `StringArray` version is used when specifing the `dtype=string`. When that happens, spark cannot serialize the column to arrow.

This sounds like an issue in Arrow side.

Reply:
Now that I think it again, adding support for the `StringArray` in arrow is the real solution, this change is just a workaround. I don't know much about the arrow internals or if they can add some metadata to support two different string types. If you feel this is better done in arrow upstream I can close the PR and investigate it from the arrow side. Otherwise, we can leave this patch adding a comment and keep investigating in arrow.

Reply:
I'll leave it to @BryanCutler (whos a maintainer for both PySpark and PyArrow).

Reply:
I'm afraid I don't think this is a right fix.

Actually this happens with `Int64`, `Int32`, or other extension dtypes.

```py
>>> spark.createDataFrame(pd.DataFrame({"A": [1, 2, 3]}, dtype="Int64"))
.../python/pyspark/sql/pandas/conversion.py:422: UserWarning: createDataFrame attempted Arrow optimization because 'spark.sql.execution.arrow.pyspark.enabled' is set to true; however, failed by the reason below:
  Cannot specify a mask or a size when passing an object that is converted with the __arrow_array__ protocol.
Attempting non-optimization as 'spark.sql.execution.arrow.pyspark.fallback.enabled' is set to true.
  warnings.warn(msg)
DataFrame[A: bigint]
```

The error message says:

```
Cannot specify a mask or a size when passing an object that is converted with the __arrow_array__ protocol.
```

, so we should follow the message.

How about at line 163,

```py
if hasattr(s.values, '__arrow_array__'):
    mask = None
else:
    mask = s.isnull()
```

Reply:
I agree with @ueshin , the correct thing to do would be to continue to use pyarrow to convert the column in this case with `mask = None`, that way the implementation is sure to do the conversion most efficiently. I think for this case it should then produce a standard arrow UTF-8 array that PySpark could handle.

Reply:
@ueshin Makes a lot of sense. I wasn't aware this was a problem for other dtypes. Removing the mask works, but writing a quick test:

```python
with self.sql_conf({"spark.sql.execution.arrow.pyspark.enabled": True}):
    pandas_df = pd.DataFrame({"col": [1, 2, 3, None]}, dtype="Int64")
    df = self.spark.createDataFrame(pandas_df)
    assert_frame_equal(pandas_df, df.toPandas())
```
fails with:
```
Attribute "dtype" are different
[left]:  Int64
[right]: float64
```

I'm going to take a look where the conversion is failing.

Reply:
I think it's fine if the `df`'s data type is `LongType` for `dtype="Int64"` because `df.toPandas()` won't keep the extension dtypes.
So the dtype of `df.toPandas()` will be `int64`, or `float64` if containing `None`.